### PR TITLE
Reset Password

### PR DESCRIPTION
**Pull request by MariamNaiga**
#21 
**What Does This PR Do?** 

- Adds a feature that allows users to reset their password

**Description of Task?**

- The user is able to reset/update their password, in the event that they have forgotten it.

**How Can You Test This Manually?**

- Send a request to `localhost:4002/api/auth/forgotPassword` that takes a body:`{"username": "***@***.com"}`
- If the user exists, the response will be an object containing a unique jwtoken, a link to the test email, and a user object.
- Send a second request to `localHost:4002/auth/user/resetPassword` that takes an authorization bearer token (the unique jwtoken from previous request) and a body `{"password":"*****"}`.
- If the password change is successful, the response will be an object containing a message of success, a link to a test confirmation email, and a user object.

**Screenshot(s)**
image
image
image
image

### CODE REVIEW

File: src/users/dto/reset-password.dto.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+export class ValidatePasswordDto {
+  @IsNotEmpty()
+  password: string;
+}
+
+export class ValidateEmailDto {
+  @IsNotEmpty()
+  @IsEmail()
+  email: string;
+}
```
Comment (lytes20) [inline src/users/dto/reset-password.dto.ts:12]: Could you please just add new lines at the end of the files

File: src/modules/security/usergroup/usergroup.model.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+export class UserGroupDto {
+    id?: string
+    name: string
+    details: string
+    roles: string[]
+    static create({name, details, roles, id}: any): UserGroupDto {
+        return {name, details, roles, id}
+    }
+}
```
Comment (lytes20) [inline src/modules/security/usergroup/usergroup.model.ts:46]: Here too

File: src/users/controllers/user.controller.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    @UseGuards(JwtAuthGuard)
+    @Put('resetPassword')
+    async resetPassword(@Body() data: ValidatePasswordDto, @Req() request: Request): Promise<ResetPasswordResponseDto> {
+      if (await (isValidPassword(data.password))) {
+        const token = request.headers.authorization.replace('Bearer ', '');
+        return this.service.resetPassword(token, data.password);
+      }
+      throw new HttpException("Invalid Password (Password Doesn't Meet Criteria)", 404);
+    }  
+}
```
Comment (lytes20) [inline src/users/controllers/user.controller.ts:36]: this file as well

File: src/users/dto/forgot-password-response.dto.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+export class ForgotPasswordResponseDto {
+  @ApiProperty()
+  token: string;
+
+  @ApiProperty()
+  mailURL: string;
+
+  @ApiProperty()
+  user: UserListDto;
+}
```
Comment (lytes20) [inline src/users/dto/forgot-password-response.dto.ts:13]: this file too

File: src/users/dto/reset-password-response.dto.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+export class ResetPasswordResponseDto {
+  @ApiProperty()
+  message: string;
+
+  @ApiProperty()
+  mailURL: string;
+
+  @ApiProperty()
+  user: UserListDto;
+}
```
Comment (lytes20) [inline src/users/dto/reset-password-response.dto.ts:13]: this file too

File: src/utils/mailerTest.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    });
+    const mailOptions = {
+        from: testAccount.user,
+        to: data.to,
+        subject: data.subject,
+        html: data.html
+    };
+    const info = await transporter.sendMail(mailOptions)
+    return nodemailer.getTestMessageUrl(info);
+}
```
Comment (lytes20) [inline src/utils/mailerTest.ts:29]: this file as well

File: src/utils/schemaHelpers.ts [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+export const configureId = (schema: Schema) => {
+    schema.virtual('id').get(function () {
+        return this._id.toHexString();
+    });
+    schema.set('toJSON', {
+        virtuals: true
+    });
+    schema.plugin(mongooseLeanVirtuals);
+    schema.plugin(mongooseHidden)
+}
```
Comment (lytes20) [inline src/utils/schemaHelpers.ts:14]: this one too

**Review by lytes20 — APPROVED** [review]

**Comment by lytes20** [comment]
LGTM 
cc @ekastimo

**Review by A23W10 — APPROVED** [review]

**Pull request merged**